### PR TITLE
[WIP] Added DatePart and DateAdd replacement samples.

### PR DESCRIPTION
**Pull request by sdanyliv**
It is sample replacement for current Sql.DatePart and Sql.DateAdd functions which implementation is sparse across SqlOptimizer's code for specific provider.

**Planned**
- [x] Replace current implementation with new functions
- [x] Remove obsolete code from SqlOptimizers
- [x] Add DateDiff functions

### CODE REVIEW

**Comment by MaceWindu** [comment]
still on radar?

**Comment by sdanyliv** [comment]
Will finsh it when found time. Not too much work left here. Just needs to ensure that this works for every provider.

**Comment by MaceWindu** [comment]
postgresql doesn't like new implementation :smile:

**Comment by sdanyliv** [comment]
Will check.

File: Tests/Linq/Linq/DateTimeFunctionsTests.cs [hunk 1]
```diff
@@ -178,6 +178,29 @@ public void DatePartSecond([DataSources] string context)
 					from t in db.Types select Sql.AsSql(Sql.DatePart(Sql.DateParts.Millisecond, t.DateTimeValue)));
 		}
 
+		[Test]
+		public void DatepartDynamicynamic(
```
Comment (MaceWindu) [inline Tests/Linq/Linq/DateTimeFunctionsTests.cs]: test name

**Comment by MaceWindu** [comment]
We decided just document `week` datepart behavior for each provider for now

**Pull request merged**